core/src/main/java/bio/terra/pearl/core/service/VersionedEntityService.java

```diff
@@ -1,9 +1,44 @@
 package bio.terra.pearl.core.service;
 
+import bio.terra.pearl.core.dao.BaseVersionedJdbiDao;
+import bio.terra.pearl.core.dao.survey.SurveyDao;
+import bio.terra.pearl.core.model.BaseEntity;
+import bio.terra.pearl.core.model.Versioned;
+import bio.terra.pearl.core.model.survey.Survey;
+import org.springframework.transaction.annotation.Transactional;
+
+import java.util.List;
 import java.util.Optional;
+import java.util.UUID;
+
+public abstract class VersionedEntityService<T extends BaseEntity & Versioned, D extends BaseVersionedJdbiDao<T>>
+        extends ImmutableEntityService<T, D> {
+
+    public VersionedEntityService(D dao) {
+        super(dao);
+    }
+    public Optional<T> findByStableId(String stableId, int version) {
+        return dao.findByStableId(stableId, version);
+    }
+
+    public List<T> findByStableId(String stableId) {
+        return dao.findByStableId(stableId);
+    }
+
+    public int getNextVersion(String stableId) {
+        return dao.getNextVersion(stableId);
+    }
 
-public interface VersionedEntityService<T> {
-    Optional<T> findByStableId(String stableId, int version);
+    @Transactional
+    public void assignPublishedVersion(UUID id) {
+        T entity = dao.find(id).get();
+        if (entity.getPublishedVersion() != null) {
+            // this survey already has a published version, do not reassign
+            return;
```

Comment:
This feels like an unexpected case. Are there sufficient checks upstream of this that we know we're saving a new version that we can make this an error? Or at least log a warning that that this is an unexpected event? In the `assignPublishedVersionIfNeeded` functions that call this, what are the situations in which this is needed vs not needed, and why wouldn't the caller know for sure?

Reply:
it's up to the caller whether this is unexpected.   Where this might occur is if OurHealth updates a survey in production, and then decides to revert back to a previous version of the survey.  when they copy irb->live to make that last change, the survey will already have a published version, and so no assignment is needed, and that's not an error.